orttraining/orttraining/python/training/pytorch_trainer_options.py

```diff
@@ -0,0 +1,376 @@
+        _internal_use.enable_internal_postprocess (bool, default is True):
+            enable internal internal post processing of the ONNX model
+        _internal_use.extra_postprocess (callable, default is None)
+            a functor to postprocess the ONNX model.
+            It does not override :py:attr:`._internal_use.enable_internal_postprocess`, but complement it
+
+    Example:
+        .. code-block:: python
+
+            opts = PytorchTrainerOptions({
+                               'batch' : {
+                                   'gradient_accumulation_steps' : 128
+                               },
+                               'device' : {
+                                   'id' : 'cuda:0',
+                                   'mem_limit' : 2*1024*1024*1024,
+                               },
+                               'lr_scheduler' : optim.lr_scheduler.LinearWarmupLRScheduler(),
+                               'mixed_precision' : {
+                                   'enabled': True,
+                                   'loss_scaler': amp.LossScaler(loss_scale=float(1 << 16))
+                               }
+            })
+            fp16_enabled = opts.mixed_precision.enabled
+     """
+
+    def __init__(self, options, _validate=True):
+        # Keep a copy of original input for debug
+        self._original_opts = dict(options)
+
+        # Add an empty dictionary for non specified nested dicts
+        subgroups = [k for k, v in _PYTORCH_TRAINER_OPTIONS_SCHEMA.items()
+                     if isinstance(v, dict) and 'type' in v and v['type'] == 'dict']
+        self._validated_opts = dict(self._original_opts)
+        if _validate:
+            for name in subgroups:
+                if name not in self._validated_opts:
+                    self._validated_opts[name] = {}
+
+            # Validates user input
+            validator = PytorchTrainerOptionsValidator(_PYTORCH_TRAINER_OPTIONS_SCHEMA)
+            self._validated_opts = validator.validated(self._validated_opts)
+            if self._validated_opts is None:
+                self._validated_opts = validator.errors
+            _validate = False
+
+        # Convert dict in object
+        for k, v in self._validated_opts.items():
+            setattr(self, k, self._wrap(v, _validate))
+
+        # Keep this in the last line
+        self._initialized = True
+
+    def __repr__(self):
+        return '{%s}' % str(', '.join("'%s': %s" % (k, repr(v))
+                                      for (k, v) in self.__dict__.items()
+                                      if k not in ['_original_opts', '_validated_opts', '_initialized']))
+
+    def __setattr__(self, k, v):
+        if hasattr(self, '_initialized'):
+            raise Exception(f"{self.__class__.__name__} is an immutable class")
+        return super().__setattr__(k, v)
+
+    def _wrap(self, v, validate):
+        if isinstance(v, (tuple, list, set, frozenset)):
```

Comment:
why for one case `validate` is propagated, but the other `False` is used?

Reply:
Same

> why for one case `validate` is propagated, but the other `False` is used?

Same as previous comment. Only during `PytorchTrainerOptions` instantiation that `_validate=True`. All other calls are `False`

Reply:
Refactored the code to fix this